src/read_write/node_iterator.rs

```diff
@@ -48,6 +52,67 @@ impl NodeIterator {
             batch_size,
         }
     }
+
+    pub fn from_data_provider<Id: ToString>(
+        data_provider: &dyn DataProvider,
+        attributes: &[&str],
+        encoding: Encoding,
+        id: &Id,
+        num_points: usize,
+        batch_size: usize,
+    ) -> Result<Self> {
+        if num_points == 0 {
+            return Ok(NodeIterator::default());
+        }
+
+        let attribute_data_types = match data_provider.meta_proto() {
+            Ok(ref meta) if meta.has_s2() => meta
+                .get_s2()
+                .get_attributes()
+                .iter()
+                .map(|a| {
+                    AttributeDataType::from_proto(a.data_type)
+                        .map(|data_type| (a.name.clone(), data_type))
+                })
+                .collect::<Result<HashMap<String, AttributeDataType>>>()?,
+            _ => vec![
+                ("color".to_string(), AttributeDataType::U8Vec3),
+                ("intensity".to_string(), AttributeDataType::F32),
```

Comment:
some octree could be just color only, will this maybe lead to problems in line 96?

Reply:
Here we just collect the possible mappings between attribute name and attribute data type, independent of the input `attributes`. In general it would be much cleaner to also store this mapping in the octree meta proto, where we could be sure about, which attributes also exist, but I think, this may be a little too much for this PR!?

Reply:
ok!